enzy_htp/structure/structure_constraint/api.py

```diff
@@ -1,43 +1,553 @@
+    type of interaction. StructureConstraints are meant to define flexible, non-package specific relationships that
+    can be translated in between different software packages.
+
     Attributes:
-        topology: Structure
-        freeze_atoms: List[Atom]
-        geom_constrain: List[Tuple[List[Atom], float]]
-
-    Derived properties:
-        structures: List[Structure]"""
-
-    def __init__(self, topology: Structure,
-                 freeze_atoms: List[Atom], geom_constrain: List[Tuple[List[Atom], float]]):
-        self._topology = topology
-        self._freeze_atoms = freeze_atoms
-        self._geom_constrain = geom_constrain
-        # init for check
-        self._is_backbone_freeze = None
-
-    def is_backbone_freeze(self) -> bool:
-        """this method saves effort when parsing to common Amber constrain"""
-        if self._is_backbone_freeze is None:
-            return set(self._topology.backbone_atoms()) == set(self._freeze_atoms)
+        atoms_: A List[Atom] containing all the atoms in the constraint.
+        target_value_ : The float() target value for the constraint.
+        params_: A Dict[str,float] with arbitrary parameters.
+    """
+
+    def __init__(self,
+                atoms:List[Atom],
+                target_value:float,
+                params:Dict[str, Any]
+                ):
+        """Ctor that exits if the supplied number of atoms is NOT correct."""
+        
+        self.atoms_ = atoms 
+        self.target_value_ = target_value 
+        self.params_ = deepcopy( params ) 
+
+        self['calc_method'] = params.get('calc_method', 'rosetta')
+
+        if not self.correct_num_atoms():
+            _LOGGER.error("Incorrect number of atoms supplied! Exiting...")
+            exit( 1 )
+
+
+    #TODO(CJ): add function that checks if topology and constraints are compatible
+    #TODO(CJ): will need to make a version of this that actually works for the ResiduePairConstraint
+
+    def clone(self) -> StructureConstraint:
+        """Clones the StructureConstraint with the original target_value."""
+        return self.__init__(atoms, self.target_value, deepcopy(self.params_))
+
+    def clone_current(self) -> StructureConstraint:
+        """Get a version of the StructureConstraint with the current value as the 
+        target value."""
+        return self.__init__(atoms, self.current_value, deepcopy(self.params_))
+
+    def is_cartesian_freeze(self) -> bool:
+        """Is this a cartesian freeze constraint?"""
+        return False
+
+    def is_distance_constraint(self) -> bool:
+        """Is this a distance constraint?"""
+        return False
+
+    def is_angle_constraint(self) -> bool:
+        """Is this an angle constraint?"""
+        return False
+
+    def is_dihedral_constraint(self) -> bool:
+        """Is this a dihedral constraint?"""
+        return False
+
+    def is_residue_pair_constraint(self) -> bool:
```

Comment:
This method should be in RosettaInterface that takes StructureConstraint() as an input arg

Reply:
Because this logic is based on the knowledge of Rosetta

Reply:
This is comparable to putting a `escort` method inside of a `cargo` class instead of `ship`, `truck`, `airplane` classes.

Reply:
I think it is fine to be here. `ResiduePairConstraint` is a type of `StructureConstraint` and is just a collection of other types of constraints.

Reply:
Oh I dont mean `ResiduePairConstraint` but the `score_energy` method should be inside of `RosettaInterface`

Reply:
Wait I see lol. I dont know why the location shifted. It must be after new commits the lines are different. It was for score_energy()

Reply:
I marked TODO but still want to see how do you see this